File: combinatorics/resource/ST/project_5/bai_toan_6/branch-and-Bound.cpp
#include <iostream>              // Thu vien nhap xuat
#include <vector>                // Thu vien vector
#include <string>                // Thu vien chuoi
#include <climits>              // Dung INT_MAX
#include <functional>           // Su dung std::function
using namespace std;

// Dinh nghia mot nut cay
struct TreeNode {
    string label;                     // Nhan cua nut
    vector<TreeNode*> children;      // Danh sach con
    TreeNode(string l) : label(l) {} // Ham khoi tao
};

int best_cost = INT_MAX;             // Bien toan cuc luu ket qua tot nhat

// Ham de quy tinh khoang cach giua hai cay
int treeEditDistanceBnB(TreeNode* t1, TreeNode* t2, int acc) {
    if (acc >= best_cost) return INT_MAX;  // Cat nhanh neu khong toi uu

    if (!t1 && !t2) {
        best_cost = min(best_cost, acc);   // Cap nhat ket qua tot nhat
        return 0;
    }
    if (!t1) {
        int cost = 1;
        for (auto c : t2->children)
            cost += treeEditDistanceBnB(nullptr, c, acc + 1);
        best_cost = min(best_cost, acc + cost);
        return cost;
    }
    if (!t2) {
        int cost = 1;
        for (auto c : t1->children)
            cost += treeEditDistanceBnB(c, nullptr, acc + 1);
        best_cost = min(best_cost, acc + cost);
        return cost;
    }

    int cost = (t1->label == t2->label) ? 0 : 1;
    int m = t1->children.size(), n = t2->children.size();
    int min_sub = INT_MAX;

    function<void(int, int, int)> dfs = [&](int i, int j, int sub_cost) {
        if (sub_cost >= min_sub) return;
        if (i == m && j == n) {
            min_sub = min(min_sub, sub_cost);
            return;
        }
        if (i < m && j < n)
            dfs(i + 1, j + 1, sub_cost + treeEditDistanceBnB(t1->children[i], t2->children[j], acc + cost + sub_cost));
        if (i < m)
            dfs(i + 1, j, sub_cost + treeEditDistanceBnB(t1->children[i], nullptr, acc + cost + sub_cost));
        if (j < n)
            dfs(i, j + 1, sub_cost + treeEditDistanceBnB(nullptr, t2->children[j], acc + cost + sub_cost));
    };

    dfs(0, 0, 0);
    int total = cost + min_sub;
    best_cost = min(best_cost, acc + total);
    return total;
}

// Ham nhap cay tu ban phim
TreeNode* inputTree() {
    string label;
    cout << "Nhap nhan goc: ";
    cin >> label;
    TreeNode* node = new TreeNode(label);
    int n;
    cout << "So con cua " << label << ": ";
    cin >> n;
    for (int i = 0; i < n; ++i) {
        cout << "-- Con thu " << i + 1 << " cua " << label << ":\n";
        node->children.push_back(inputTree());
    }
    return node;
}

int main() {
    cout << "Nhap cay 1:\n";
    TreeNode* t1 = inputTree();
    cout << "Nhap cay 2:\n";
    TreeNode* t2 = inputTree();
    cout << "Edit distance (BnB): " << treeEditDistanceBnB(t1, t2, 0) << endl;
    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_6/branch-and-Bound.py
# Lop cay voi nhan va danh sach con
class TreeNode:
    def __init__(self, label):
        self.label = label                 # Luu nhan cua nut
        self.children = []                # Khoi tao danh sach con rong

# Ham tinh khoang cach giua hai cay bang Branch-and-Bound
def tree_edit_distance_bnb(t1, t2):
    best_cost = [float('inf')]  # Chi phi nho nhat tim duoc

    def recurse(t1, t2, acc):
        if acc >= best_cost[0]:
            return float('inf')
        if t1 is None and t2 is None:
            best_cost[0] = min(best_cost[0], acc)
            return 0
        if t1 is None:
            cost = 1 + sum(recurse(None, c, acc+1) for c in t2.children)
            best_cost[0] = min(best_cost[0], acc + cost)
            return cost
        if t2 is None:
            cost = 1 + sum(recurse(c, None, acc+1) for c in t1.children)
            best_cost[0] = min(best_cost[0], acc + cost)
            return cost

        cost = 0 if t1.label == t2.label else 1
        m, n = len(t1.children), len(t2.children)
        min_sub = float('inf')

        def dfs(i, j, sub_cost):
            nonlocal min_sub
            if sub_cost >= min_sub: return
            if i == m and j == n:
                min_sub = min(min_sub, sub_cost)
                return
            if i < m and j < n:
                dfs(i+1, j+1, sub_cost + recurse(t1.children[i], t2.children[j], acc + cost + sub_cost))
            if i < m:
                dfs(i+1, j, sub_cost + recurse(t1.children[i], None, acc + cost + sub_cost))
            if j < n:
                dfs(i, j+1, sub_cost + recurse(None, t2.children[j], acc + cost + sub_cost))

        dfs(0, 0, 0)
        total = cost + min_sub
        best_cost[0] = min(best_cost[0], acc + total)
        return total

    return recurse(t1, t2, 0)

# Ham nhap cay tu ban phim
def input_tree():
    label = input("Nhap nhan goc: ")
    node = TreeNode(label)
    n = int(input(f"So con cua {label}: "))
    for i in range(n):
        print(f"-- Con thu {i+1} cua {label}:")
        child = input_tree()
        node.children.append(child)
    return node

if __name__ == "__main__":
    print("Nhap cay 1:")
    t1 = input_tree()
    print("Nhap cay 2:")
    t2 = input_tree()
    print("Edit distance (BnB):", tree_edit_distance_bnb(t1, t2))